### DISCUSSION TITLE
image upload failed

### DISCUSSION THREAD

**Discussion by Makiny** [outline/outline#3932] [Self Hosting] [answered]
**Expected behavior**
When upload image in outline, notice "Upload failed".

S3 service is minio, the imfomation is :
<img width="486" alt="image" src="https://user-images.githubusercontent.com/28851690/183237301-8aec0af4-ab66-49b7-a8ff-1cf99a5a2771.png">

outline config is :
is already created Buckets in minio, Buckets name is "outline", the outline and minio at the same network.

<img width="1092" alt="image" src="https://user-images.githubusercontent.com/28851690/183237343-a58c6815-622d-493c-8579-857736e54814.png">

**Screenshots**
<img width="407" alt="image" src="https://user-images.githubusercontent.com/28851690/183237175-f068a56c-618f-4f07-a253-a8fff2e6c157.png">

**Outline 
- Install: self hosted(docker)
- Version: 0.65.2

**Desktop 
 - OS: macos
 - Browser safari
 - Version 15.6 (17613.3.9.1.5)

**Comment by Makiny** [comment, marked as answer]
I solved this problem, this is an error caused by Unknown sub-system, just configure the api address of minio to https

**Comment by FrankGiesecke** [comment]
Please, can you explain something more? I run into the same error using the SysEleven  Object Storage.

Sadly to say that I didn't find any useful outline log.

**Reply by Makiny** [reply]
You can open the browser's debug page (F12 for chrome in windows), then upload the image in the outline, and check what error is returned by the api for uploading the image, like this (of course, mine did not report an error):

<img width="1506" alt="image" src="https://user-images.githubusercontent.com/28851690/187661778-0b172b9b-a295-413b-b141-995aa9b2239d.png">

**Comment by FrankGiesecke** [comment]
@Makiny 
if I do this, I see an 403 Error.

Request Headers:
```
:authority: xxxxxxxx
:method: POST
:path: /s3/
:scheme: https
accept: */*
accept-encoding: gzip, deflate, br
accept-language: de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7
content-length: 201669
content-type: multipart/form-data; boundary=----WebKitFormBoundaryMH8JNZtOTrzHDAQQ
cookie: lastSignedIn=email; accessToken=exxxxxxx
dnt: 1
origin: https://xxxxxxxx
sec-ch-ua: "Chromium";v="104", " Not A;Brand";v="99", "Google Chrome";v="104"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "macOS"
sec-fetch-dest: empty
sec-fetch-mode: cors
sec-fetch-site: same-origin
user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36
```

Payload:
```
Cache-Control: max-age=31557600
Content-Type: image/jpeg
Content-Disposition: attachment
key: public/xxxx/yyyy/foto.jpeg
acl: public-read
bucket: bucket
X-Amz-Algorithm: AWS4-HMAC-SHA256
X-Amz-Credential: 9c340ffb624547bea74105985a45e953/20220831/cbk/s3/aws4_request
X-Amz-Date: 20220831T110811Z
Policy: xxx
X-Amz-Signature: xxx
file: (binary)
```

Response Body:
```
<Error>
<Code>
AccessDenied
</Code>
<Message>
Access Denied
</Message>
<Resource>
/bucket/public/xxxx/yyyy/foto.jpeg
</Resource>
<RequestId>
6502007073693672726
</RequestId>
</Error>
```

I can upload files using `s3cmd` and the credentials I've configured in outline.

**Reply by tommoor** [reply]
Perhaps your storage service has more details on why permission was denied. Is `public-read` allowed?

**Reply by AndreaPersico** [reply]
Dear all,
I have a very similar issue that I'm not able to solve.
My installation is based upon Docker with NGINX reverse proxy, Keycloak auth, MinIO: I've almost finished the configuration in my homelab but uploading images is still not possible (I've already checked the other posts without success)
The error message on Outline is very generic and if I dig into Chrome Developer tools I see a strange result on the bucket post request.
![image](https://user-images.githubusercontent.com/49158551/189770678-e8d0740a-6f0b-4d59-9afd-ea22bb2e020c.png)

**Outline
Install: self hosted(docker)
Version: 0.65.2

**Outline Logs
```
{"name":"views.create","modelId":"a94eea0d-9980-400a-b7f0-969b2d408954","attempt":0,"label":"worker","level":"info","message":"Processing views.create"}
{"userId":"0eeaf1b2-2623-4f12-87cf-007876c342ce","label":"hocuspocus","level":"info","message":"Loaded document \"document.e27af7e2-4c77-49bd-bed8-07b5308ed1a0\""}
{"name":"attachments.create","modelId":null,"label":"worker","level":"info","message":"WebhookProcessor running attachments.create"}
```

Any idea?
Are there any updated guide for configuring Outline and MinIO? I've found some outdated tutorials and it is a pity that a promising wiki in my case is so frustrating to install and configure on a self-hosted environment.
Thanks for your help!

**Reply by FrankGiesecke** [reply]
Further investigation shows, that there is no access token in the request header send by outline.
So the S3 is not able to authenticate the request.

**Reply by tommoor** [reply]
Access tokens are not used for S3. Either the resource should be public, like the one you posted at the start of the thread or private files have a signed url.

**Reply by axelsimon** [reply]
@tommoor Could you explain what  the access key and secret access key are for then?
And for private buckets, how do URLs get a valid signature? Is this something Outline does automatically?
